Add virtio-iommu support

sboeuf opened the pull request:
This pull request intends to bring support for virtual IOMMU relying on the brand new `virtio-iommu` device specification.

Fixes #314

sboeuf commented:
@rbradford @sameo 
Just wanted to give some update here. I've been able, based on the patches from @eauger and @jpbrucker, to boot a VM with a `virtio-iommu-pci` device along with a paravirtualized IOMMU and a PCI root complex IORT nodes. Altogether, this exposes to the guest some specific devices that are described in the PCI root complex node, as attached to the IOMMU. That's how we can see some devices being part of iommu groups under `/sys/kernel/iommu_groups/`.
Because of this, I could use VFIO to bind the device that I had identified as being attached to the IOMMU, and then start a nested VM from this first VM.
And here are the two use cases, either we're passing a virtio device, or we're passing a real device that has been already passed through.
This implies a lot of inter dependency between virtio devices, VFIO devices and the virtio-iommu device.
- In case of virtio, we need to implement the `VIRTIO_F_IOMMU_PLATFORM` feature, which implies to translate addresses coming from the guest driver before to try reading from the guest memory. Indeed, if we virtually attach a virtio device to a vIOMMU, we cannot assume anymore that the guest address is a physical address. It will be an IOVA, which is why we need to handle it from any virtio device that would support this feature. That means we need to store the mapping provided by the `virtio-iommu` device, and use it to perform the translation.
- In case of VFIO, we want to call `VFIO_DMA_MAP` ioctl on the existing VFIO container, which will update the previous mapping in the physical IOMMU. This case is different as every time we receive a new MAP request through virtio-iommu, we need to trigger a VFIO ioctl call.

There's one last thing that is going to be challenging to tackle, we need to know the BDF of each device we're placing on the PCI bus, as we need this to provide it to the PCI root complex node, so that it can identify which devices are attached to the IOMMU. I think we need to build a list of the devices that will end up attached to the IOMMU, and once all devices have been initialized, we can generate the proper IORT table with the associated IORT node and mappings.

I wanted to throw my thoughts here as we need to have some design discussions on how to handle this :)

rbradford commented:
> And here are the two use cases, either we're passing a virtio device, or we're passing a real device that has been already passed through.

Are you talking about nested VFIO here?  Is that a very frequent use case?

sameo commented:
On Fri, Sep 27, 2019 at 06:12:45PM -0700, Sebastien Boeuf wrote:
> This implies a lot of inter dependency between virtio devices, VFIO devices and the virtio-iommu device.
> - In case of virtio, we need to implement the `VIRTIO_F_IOMMU_PLATFORM` feature, which implies to translate addresses coming from the guest driver before to try reading from the guest memory. Indeed, if we virtually attach a virtio device to a vIOMMU, we cannot assume anymore that the guest address is a physical address. It will be an IOVA, which is why we need to handle it from any virtio device that would support this feature. That means we need to store the mapping provided by the `virtio-iommu` device, and use it to perform the translation.

I guess this is the price we have to pay for pretending to be a physical
device. Who exposes the `VIRTIO_F_IOMMU_PLATFORM` feature? I hope we can
decide if we need to do the translation at runtime or not. In case of
non directly assigned virtio devices, supporting virtio-iommu should
have zero impact.

> - In case of VFIO, we want to call `VFIO_DMA_MAP` ioctl on the existing VFIO container, which will update the previous mapping in the physical IOMMU. This case is different as every time we receive a new MAP request through virtio-iommu, we need to trigger a VFIO ioctl call.
>

I suspect this is not something that happens *that* often for actual
phyical devices, when they are managed by a sane guest driver.

> There's one last thing that is going to be challenging to tackle, we need to know the BDF of each device we're placing on the PCI bus, as we need this to provide it to the PCI root complex node, so that it can identify which devices are attached to the IOMMU. I think we need to build a list of the devices that will end up attached to the IOMMU, and once all devices have been initialized, we can generate the proper IORT table with the associated IORT node and mappings.
>

Hmm...Right now we have that knowledge, but I guess that would prevent
us from doing vfio+hotplug, i.e. directly assigning a PCI device while
the VM is running?

sboeuf commented:
@rbradford 
>Are you talking about nested VFIO here? Is that a very frequent use case?

Well VFIO nested is only one subsequent application out of the more general use case. If we expose to the guest a virtio device that we attach to this vIOMMU, then we need to handle the IOVA->PA translation. And if we expose a real device that we passed through, and that we attach this device to the vIOMMU, then the driver (let's say e1000 in case we're dealing with a NIC) will start doing some IOVA allocations, which need to be known by our VMM. So those cases don't even involve nesting anything, it's just a case where the guest drivers behave differently because they can see the devices are attached to an IOMMU.

sboeuf commented:
@sameo 
>I guess this is the price we have to pay for pretending to be a physical
device. Who exposes the `VIRTIO_F_IOMMU_PLATFORM` feature? I hope we can
decide if we need to do the translation at runtime or not. In case of
non directly assigned virtio devices, supporting virtio-iommu should
have zero impact.

Yes of course we can decide to expose the feature or not. But the point is to be able to expose the feature for any virtio device that we would attach to this vIOMMU.

>I suspect this is not something that happens *that* often for actual
phyical devices, when they are managed by a sane guest driver.

In case of VFIO, there are two stages. When the `e1000` driver is being loaded from the guest, there might be a bunch of random allocations. But once you unbind the device from the e1000 driver and bind it to VFIO, once the nested VMM instance uses VFIO, the entire guest L2 memory is pinned, which is why we get a bunch of 4k pages allocations for the entire range, but this happens only once.

>Hmm...Right now we have that knowledge, but I guess that would prevent
us from doing vfio+hotplug, i.e. directly assigning a PCI device while
the VM is running?

Oh we do have that knowledge?! That's great news then. Why do you think this might interfere with hotplug? In the case of vIOMMU, we don't need a specific device slot, we only need to anticipate where the device is going to end up from a PCI hierarchy perspective, that's it. If we already have a way to know this, then we should be fine with this issue.

sboeuf commented:
@rbradford @sameo This PR is now ready for a round of review. I cleaned up the commits and added a very simple integration test. This has support for both `virtio-blk` and `virtio-net`.

The next steps are:
- Having a reliable kernel branch with the patches (waiting for virtio-iommu maintainers to get some updates)
- Based on this new kernel, add some tests where using `virtio-blk` and `virtio-net` with the option `iommu=on`
- Update the VFIO integration test, replacing QEMU with cloud-hypervisor 
- Implement iommu support for all remaining virtio devices
- Last but not the least, add the virtual iommu support to VFIO devices

Let me know how you want to proceed with those steps. Should we merge something first and come up with follow up PRs for all those bullet points?

rbradford commented:
Please make sure the README is updated with the command line changes and the purpose of the feature and any impact turning it on might have (performance) and the reason why you would want to turn it.

rbradford commented on vm-virtio/src/iommu.rs:
```diff
@@ -0,0 +1,820 @@
+
+// Default to 4KiB page size.
+#[allow(unused)]
+const VIRTIO_IOMMU_PAGE_SIZE_MASK: u64 = 1 << 12;
```
Personally, I prefer `4 << 10`

rbradford commented on vmm/src/device_manager.rs:
```diff
@@ -383,8 +392,25 @@ impl DeviceManager {
+            Some(mapping)
+        } else {
+            None
+        };
```
Could you move this down into the `pci_support` block below? Then you could get away without naming the variable with the _ prefix.

sboeuf replied:
Yes good point!

rbradford commented on vmm/src/device_manager.rs:
```diff
@@ -563,22 +606,30 @@ impl DeviceManager {
+                            raw_img,
+                            disk_cfg.path.clone(),
+                            false,
+                            disk_cfg.iommu,
```
What happens if a user turns it on for a disk and forgets to turn it on globally?

sboeuf replied:
The virtio feature flag is enabled in this case, but because the virtio-iommu is not created, it has strictly no impact on the way the device works, with no impact on performance.

rbradford commented:
Overall lgtm. I see no harm in merging a WIP, this PR does clash with #273

sboeuf commented:
@sameo @rbradford All comments addressed, and the PR is already rebased on #273, so it's ready to be merged whenever #273 gets merged.

sameo commented:
Rebased on top of master.

sameo commented on docs/iommu.md:
```diff
@@ -0,0 +1,87 @@
+    --disk path=clear-kvm.img,iommu=on \
+    --kernel custom-bzImage \
+    --cmdline "console=ttyS0 root=/dev/vda3" \
+    --iommu
```
Could we have a simpler user experience by implicitly enabling the virtio IOMMU if any of the devices (parvirt'ed or directly assigned) asks for going through the IOMMU.
The `--iommu` case would still make sense for when we'll want to do device hotplug and have it go through our IOMMU, although I guess we might also dynamically add it.
Bottom line is: can we make the `--iommu` cli switch optional?

sboeuf replied:
Yes you're right, I changed the code to enable virtio-iommu creation if at least one of the devices has `iommu=on` flag.

sameo commented on docs/iommu.md:
```diff
@@ -0,0 +1,87 @@
+for updating the physical DMA remapping table. This will happen through the
+VFIO framework, and will eventually allow for passing a physical device through
+multiple layers of virtualization.
+
```
Another question that this documentation triggers: if I do a device passthrough with `--iommu`, I assume the guest driver will switch to using IOVA and implictly expect our virtual IOMMU to do the IOVA->PA translation. I see that this PR passes the VFIO CI tests, though: Any idea how that works?

sboeuf replied:
Because we have to choose which devices are attached to the vIOMMU. By default, a device is not attached to it, therefore the driver does not use IOVAs.

sboeuf commented on docs/iommu.md:
```diff
@@ -0,0 +1,87 @@
+access the guest memory, the same way a physical IOMMU would prevent a
+physical device from accessing the physical memory. The security gain is only
+visible from a guest perspective, which does not apply if the guest we're
+taking as example is corrupted. The benefits about security can be seen in
```
Yes `malicious` is the right word.

sboeuf commented on docs/iommu.md:
```diff
@@ -0,0 +1,87 @@
+visible from a guest perspective, which does not apply if the guest we're
+taking as example is corrupted. The benefits about security can be seen in
+case of nested virtualization as it can protect the first layer guest from
+malicious guests running in L2. 
```
Yes let me rephrase all this properly.

sboeuf commented:
@sameo All comments addressed, PTAL
